[misc] Update json importability processor with requiredSubjectTypes support

acrowthe opened the pull request:
Add the ability for JsonProcessors to rename a property
Add custom handling to Importability for requiredSubjectTypes properties:
- Replace the subject type reference with the subject type path
- Add a prefix to the property name so the json to xml script knows it is a reference

acrowthe committed:
[misc] Update json importability processor with requiredSubjectTypes support

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/ImportableProcessor.java:
```diff
@@ -94,4 +105,17 @@ public JsonValue processChild(final Node node, final Node child, final JsonValue
+    public String processPropertyName(final Node node, final Property property, final String input)
+    {
+        try {
+            if (SUBJECT_TYPES_PROPERTY.equals(property.getName())) {
+                return "jcr:reference:" + SUBJECT_TYPES_PROPERTY;
+            }
```
I think this should be generic, you can see the type of the property, and if it's reference use the prefix.


```suggestion
            if (property.getType() == PropertyType.REFERENCE) {
                return "jcr:reference:" + property.getName();
            }
```

Also needs an import for `PropertyType`.

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/spi/ResourceJsonProcessor.java:
```diff
@@ -154,6 +154,24 @@ default JsonValue processProperty(final Node node, final Property property, fina
+     * {@code null}, then the property is skipped from the parent node's serialization. Only the outcome of the last
+     * processor is taken into consideration when serializing the parent node. The default implementation simply
+     * returns the input unmodified
+     * @param node the node who's properties name is being deternmined
```
```suggestion
     * @param node the node whose properties name is being determined
```

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/ImportableProcessor.java:
```diff
@@ -69,8 +74,14 @@ public JsonValue processProperty(final Node node, final Property property, final
             } else if (propertyName.startsWith("sling:")) {
                 // Remove all sling properties
                 result = null;
+            } else if (SUBJECT_TYPES_PROPERTY.equals(propertyName)) {
```
I think this could also benefit from a check on `type == reference`. This way the processor makes other node types, not just questionnaires, importable.

acrowthe committed:
[misc] Update json importability processor with requiredSubjectTypes support
- Improve requireSubjectTypes handling to be generic to all references
- Fix a couple typos in javadoc

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/internal/ImportableProcessor.java:
```diff
@@ -74,13 +73,19 @@ public JsonValue processProperty(final Node node, final Property property, final
+                    }
+                    result = arrayBuilder.build();
+                } else {
+                    Node referencedNode = property.getSession().getNodeByIdentifier(property.getValue().getString());
```
There's `property.getNode()`, which returns the node referenced by the property.

acrowthe replied:
Ah, I remembered thinking there should be a `.getNode()` but I think I was trying to use `value.getNode()` which doesn't exist.

acrowthe committed:
[misc] Update json importability processor with requiredSubjectTypes support
- Small code cleanup

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/spi/ResourceJsonProcessor.java:
```diff
@@ -154,6 +154,24 @@ default JsonValue processProperty(final Node node, final Property property, fina
 
+    /**
+     * Called for each property of a serialized node, allows adjusting the name of the JSON serialized property. The
+     * first processer invoked will recieve the properties original name as the input string, and each subsequent
```
```suggestion
     * first processor invoked will receive the properties original name as the input string, and each subsequent
```

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/spi/ResourceJsonProcessor.java:
```diff
@@ -154,6 +154,24 @@ default JsonValue processProperty(final Node node, final Property property, fina
+    /**
+     * Called for each property of a serialized node, allows adjusting the name of the JSON serialized property. The
+     * first processer invoked will recieve the properties original name as the input string, and each subsequent
+     * processor being invoked recieves the output of the previous one. If at the end of the process the outcome is
```
```suggestion
     * processor being invoked receives the output of the previous one. If at the end of the process the outcome is
```

sdumitriu commented on modules/utils/src/main/java/io/uhndata/cards/serialize/spi/ResourceJsonProcessor.java:
```diff
@@ -154,6 +154,24 @@ default JsonValue processProperty(final Node node, final Property property, fina
+     * {@code null}, then the property is skipped from the parent node's serialization. Only the outcome of the last
+     * processor is taken into consideration when serializing the parent node. The default implementation simply
+     * returns the input unmodified
+     * @param node the node whose properties name is being determined
```
```suggestion
     * @param node the node whose property name is being determined
```

sashaandjic commented:
How can I test this PR?

marta- commented:
Export a questionnaire with requiredSubjectTypes from cards, run it through the scripts for reimport, check that requiredSubjectTypes is in the right format in the XML.

sdumitriu committed:
[misc] Typo fixes
